core/src/main/java/org/infinispan/protostream/impl/TagReaderImpl.java

```diff
@@ -1066,186 +995,75 @@ InputStream getInputStream() {
-      private byte[] readRawBytesLarge(int length) throws IOException {
-         if (length < 0) {
-            throw new IllegalArgumentException("Length must not be negative");
-         }
-
-         int total = bytesBeforeStart + pos + length;
-         if (total - globalLimit > 0) {
-            throw log.globalLimitExceeded();
-         }
-         if (total > limit) {
-            // limit exceeded, skip up to limit and fail
-            skipRawBytes(limit - bytesBeforeStart - pos);
-            throw log.messageTruncated();
-         }
-
-         int buffered = end - pos;
-         int needed = length - buffered;
-         if (needed <= 0) {
-            throw new IllegalStateException("The needed data already exists in buffer!");
-         }
-         int oldPos = pos;
-         bytesBeforeStart += end;
-         pos = 0;
-         end = 0;
-
-         if (needed < ProtobufUtil.DEFAULT_STREAM_BUFFER_SIZE || needed <= in.available()) {
-            byte[] bytes = new byte[length];
-            System.arraycopy(buf, oldPos, bytes, 0, buffered);
-            while (buffered < bytes.length) {
-               int read = in.read(bytes, buffered, length - buffered);
-               if (read <= 0) {
-                  throw log.messageTruncated();
-               }
-               bytesBeforeStart += read;
-               buffered += read;
-            }
-            return bytes;
-         }
-
-         // read in segments to avoid allocating full length at once to prevent sudden death
-         List<byte[]> segments = new ArrayList<>();
-         while (needed > 0) {
-            byte[] segment = new byte[Math.min(needed, ProtobufUtil.DEFAULT_STREAM_BUFFER_SIZE)];
-            int segPos = 0;
-            while (segPos < segment.length) {
-               int read = in.read(segment, segPos, segment.length - segPos);
-               if (read <= 0) {
-                  throw log.messageTruncated();
-               }
-               segPos += read;
-               bytesBeforeStart += read;
-            }
-            segments.add(segment);
-            needed -= segment.length;
-         }
-
-         // stitch the segments and hope not to blow up
-         byte[] bytes = new byte[length];
-         System.arraycopy(buf, oldPos, bytes, 0, buffered);
-         int segPos = buffered;
-         for (int i = 0; i < segments.size(); i++) {
-            byte[] segment = segments.get(i);
-            System.arraycopy(segment, 0, bytes, segPos, segment.length);
-            segPos += segment.length;
-            segments.set(i, null);
-         }
-         return bytes;
+         throw log.messageTruncated();
       }
 
       @Override
       protected void skipRawBytes(int length) throws IOException {
-         if (length <= end - pos && length >= 0) {
+         if (length <= limit - pos && length >= 0) {
             pos += length;
+            long skipAmount = in.skip(length);
```

Comment:
same question here
```suggestion
            long skipAmount = in.skipNBytes(length);
```

Reply:
Oh the reason I couldn't use this is because it requires Java 12 or newer

Reply:
ah ah, I checked the JDK version of `readNBytes` and I assumed this one was the same :rofl: